Emission Summary Count Off-By-One Error

Hello all,

## Incorrect Summary Count

I was running `cargo run --example summaries` on `421fd76e`  and I noticed that the summary miscounts the number of suppressed messages.

```text
=== Suppression Summary Example ===

This example demonstrates how summaries show WHAT events are suppressed,
not just signature hashes. You'll see clear, human-readable descriptions.

Human-Readable Suppression Summaries
-------------------------------------

Configuration:
  - Policy: count-based (first 3 allowed)
  - Summary interval: 1 seconds
  - Formatter: default (WARN level)

Scenario: High-traffic application with repetitive events

📝 Emitting 20 identical user login events:
2026-03-06T16:54:53.911193Z  INFO summaries: User login successful user_id="user_123"
2026-03-06T16:54:53.911469Z  INFO summaries: User login successful user_id="user_123"
2026-03-06T16:54:53.911739Z  INFO summaries: User login successful user_id="user_123"
   → First 3 allowed, rest suppressed

❌ Emitting 15 identical database error events:
2026-03-06T16:54:53.913360Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
2026-03-06T16:54:53.913651Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
2026-03-06T16:54:53.913949Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
   → First 3 allowed, rest suppressed

⚠️  Emitting 10 identical cache miss warnings:
2026-03-06T16:54:53.914939Z  WARN summaries: Cache miss detected cache_key="session:abc"
2026-03-06T16:54:53.915155Z  WARN summaries: Cache miss detected cache_key="session:abc"
2026-03-06T16:54:53.915367Z  WARN summaries: Cache miss detected cache_key="session:abc"
   → First 3 allowed, rest suppressed

⏳ Waiting for summary emission (1.1 seconds)...

🔍 Watch the summaries below - notice how they tell you EXACTLY
   what was suppressed, not just a meaningless hash!

2026-03-06T16:54:54.811095Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=ad6c96dc5f891806 count=8
2026-03-06T16:54:54.811549Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=235ee95dc827a405 count=18
2026-03-06T16:54:54.811892Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=379e76651e8678a0 count=13
```

Notice how the counts on the final warnings are 8, 18, and 13. I posit that they should instead be 7, 17, and 12, since there were originally 10, 20, and 15 events emitted, and the suppression count policy is to allow 3 and suppress the rest.

## Recursive Summaries

On another note, something interesting happens if you continue logging the summaries.

```text
⏳ Waiting for summary emission (1.1 seconds)...

🔍 Watch the summaries below - notice how they tell you EXACTLY
   what was suppressed, not just a meaningless hash!

2026-03-06T16:57:22.729575Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=207bc98ec82a7ea0 count=8
2026-03-06T16:57:22.730021Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=530b5663cbddf9bb count=18
2026-03-06T16:57:22.730351Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=9f98f6a63c9b9c70 count=13

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T16:57:23.729358Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=034d6f700fe47399 count=1
2026-03-06T16:57:23.729551Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=8d79b748263f6833 count=1
2026-03-06T16:57:23.729712Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=207bc98ec82a7ea0 count=8
2026-03-06T16:57:23.729907Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=b2e58c9091ec465c count=1
2026-03-06T16:57:23.730096Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=530b5663cbddf9bb count=18
2026-03-06T16:57:23.730299Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=9f98f6a63c9b9c70 count=13

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T16:57:24.730258Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=034d6f700fe47399 count=1
2026-03-06T16:57:24.730565Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=8d79b748263f6833 count=1
2026-03-06T16:57:24.730904Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=ddfc1e180a07612c count=1
2026-03-06T16:57:24.731225Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=d953a1517c4bf39a count=1
2026-03-06T16:57:24.731770Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=207bc98ec82a7ea0 count=8
2026-03-06T16:57:24.732175Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=baeefdc8727e6c02 count=1
2026-03-06T16:57:24.732492Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=b2e58c9091ec465c count=1
2026-03-06T16:57:24.732658Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=530b5663cbddf9bb count=18
2026-03-06T16:57:24.732968Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=9f98f6a63c9b9c70 count=13

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T16:57:25.730349Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=034d6f700fe47399 count=1
2026-03-06T16:57:25.730727Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=8d79b748263f6833 count=1
2026-03-06T16:57:25.731041Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=ddfc1e180a07612c count=1
2026-03-06T16:57:25.731562Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=4c4c38844a3fdbc9 count=1
2026-03-06T16:57:25.731919Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=d953a1517c4bf39a count=1
2026-03-06T16:57:25.732327Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=baeefdc8727e6c02 count=1
2026-03-06T16:57:25.732657Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=b2e58c9091ec465c count=1
2026-03-06T16:57:25.733248Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=52f75388ad53aa92 count=1
2026-03-06T16:57:25.733731Z  WARN tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 1 times over 0.00s: [WARN] tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=7986bc2e6ef75f0c count=1
```

The summary messages themselves start getting caught and duplicating themselves infinitely. Is that the expected behavior? As a second note, this behavior also shows that the "suppressed 1 times" message is incorrect, because the summary messages were never actually suppressed.

## Duplicated Summaries Fix

I ended up fixing this by using the `.with_exempt_targets` method on the throttle layer. 

```rust
    let rate_limit_layer = TracingRateLimitLayer::builder()
        .with_policy(Policy::count_based(3).unwrap())
        .with_active_emission(true)
        .with_summary_interval(Duration::from_secs(1))
        .with_exempt_targets("tracing_throttle::infrastructure::layer")
        .build()
        .unwrap();
```


```text
=== Suppression Summary Example ===

This example demonstrates how summaries show WHAT events are suppressed,
not just signature hashes. You'll see clear, human-readable descriptions.

Human-Readable Suppression Summaries
-------------------------------------

Configuration:
  - Policy: count-based (first 3 allowed)
  - Summary interval: 1 seconds
  - Formatter: default (WARN level)

Scenario: High-traffic application with repetitive events

📝 Emitting 20 identical user login events:
2026-03-06T17:06:28.076232Z  INFO summaries: User login successful user_id="user_123"
2026-03-06T17:06:28.076505Z  INFO summaries: User login successful user_id="user_123"
2026-03-06T17:06:28.076779Z  INFO summaries: User login successful user_id="user_123"
   → First 3 allowed, rest suppressed

❌ Emitting 15 identical database error events:
2026-03-06T17:06:28.077875Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
2026-03-06T17:06:28.078127Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
2026-03-06T17:06:28.078455Z ERROR summaries: Database connection timeout error_code="DB_CONN_FAILED"
   → First 3 allowed, rest suppressed

⚠️  Emitting 10 identical cache miss warnings:
2026-03-06T17:06:28.079390Z  WARN summaries: Cache miss detected cache_key="session:abc"
2026-03-06T17:06:28.079525Z  WARN summaries: Cache miss detected cache_key="session:abc"
2026-03-06T17:06:28.079729Z  WARN summaries: Cache miss detected cache_key="session:abc"
   → First 3 allowed, rest suppressed

⏳ Waiting for summary emission (1.1 seconds)...

🔍 Watch the summaries below - notice how they tell you EXACTLY
   what was suppressed, not just a meaningless hash!

2026-03-06T17:06:28.973725Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=83fb4f9dc08444e2 count=13
2026-03-06T17:06:28.974069Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=bcb46ce3fe3f3505 count=8
2026-03-06T17:06:28.974357Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=72f00ee08a88bfe2 count=18

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T17:06:29.974568Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=83fb4f9dc08444e2 count=13
2026-03-06T17:06:29.974857Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=bcb46ce3fe3f3505 count=8
2026-03-06T17:06:29.974943Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=72f00ee08a88bfe2 count=18

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T17:06:30.973524Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=83fb4f9dc08444e2 count=13
2026-03-06T17:06:30.973804Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=bcb46ce3fe3f3505 count=8
2026-03-06T17:06:30.973883Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=72f00ee08a88bfe2 count=18

⏳ Waiting for summary emission (1.1 seconds)...

2026-03-06T17:06:31.974198Z  WARN tracing_throttle::infrastructure::layer: Suppressed 13 times over 0.00s: [ERROR] summaries: Database connection timeout signature=83fb4f9dc08444e2 count=13
2026-03-06T17:06:31.974499Z  WARN tracing_throttle::infrastructure::layer: Suppressed 8 times over 0.00s: [WARN] summaries: Cache miss detected signature=bcb46ce3fe3f3505 count=8
2026-03-06T17:06:31.974588Z  WARN tracing_throttle::infrastructure::layer: Suppressed 18 times over 0.00s: [INFO] summaries: User login successful signature=72f00ee08a88bfe2 count=18

=== Example Complete ===
```

Is it expected for users to use that to fix the summary emissions?


## Conclusion

Let me know if I'm missing anything obvious, and thank you for your time.

--- a/src/application/emitter.rs
+++ b/src/application/emitter.rs
@@ -482,9 +482,9 @@ mod tests {
 
         // Verify counts
         let counts: Vec<usize> = summaries.iter().map(|s| s.count).collect();
-        assert!(counts.contains(&6)); // 5 + 1 (initial)
-        assert!(counts.contains(&11)); // 10 + 1
-        assert!(counts.contains(&16)); // 15 + 1
+        assert!(counts.contains(&5));
+        assert!(counts.contains(&10));
+        assert!(counts.contains(&15));
     }
 
     #[test]
@@ -516,7 +516,7 @@ mod tests {
 
         // Only the high-count event should be included
         assert_eq!(summaries.len(), 1);
-        assert_eq!(summaries[0].count, 15); // 14 + 1 initial
+        assert_eq!(summaries[0].count, 14);
     }
 
     #[cfg(feature = "async")]

--- a/src/application/limiter.rs
+++ b/src/application/limiter.rs
@@ -306,9 +306,9 @@ mod tests {
         assert_eq!(limiter.check_event(sig), LimitDecision::Suppress);
         assert_eq!(limiter.check_event(sig), LimitDecision::Suppress);
 
-        // Check counter - initial creation counts as 1, plus 3 recorded = 4 total
+        // Check counter - 3 suppressions recorded
         registry.with_event_state(sig, |state, _now| {
-            assert_eq!(state.counter.count(), 4);
+            assert_eq!(state.counter.count(), 3);
         });
     }
 
@@ -608,7 +608,7 @@ mod tests {
 
         // Verify state exists
         let initial_count = registry.with_event_state(sig, |state, _| state.counter.count());
-        assert_eq!(initial_count, 1);
+        assert_eq!(initial_count, 0);
 
         // Open circuit breaker
         for _ in 0..10 {

--- a/src/application/registry.rs
+++ b/src/application/registry.rs
@@ -200,7 +200,7 @@ mod tests {
 
         // First access creates state
         registry.with_event_state(sig, |state, _now| {
-            assert_eq!(state.counter.count(), 1);
+            assert_eq!(state.counter.count(), 0);
         });
 
         assert_eq!(registry.len(), 1);

--- a/src/domain/summary.rs
+++ b/src/domain/summary.rs
@@ -35,11 +35,11 @@ impl Clone for SuppressionCounter {
 }
 
 impl SuppressionCounter {
-    /// Create a new counter with initial suppression.
+    /// Create a new counter (initially zero suppressions).
     pub fn new(initial_timestamp: Instant) -> Self {
         let nanos = Self::instant_to_nanos(initial_timestamp);
         Self {
-            suppressed_count: AtomicUsize::new(1),
+            suppressed_count: AtomicUsize::new(0),
             first_suppressed_nanos: AtomicU64::new(nanos),
             last_suppressed_nanos: AtomicU64::new(nanos),
         }
@@ -269,13 +269,13 @@ mod tests {
         let now = Instant::now();
         let counter = SuppressionCounter::new(now);
 
+        assert_eq!(counter.count(), 0);
+
+        counter.record_suppression(now);
         assert_eq!(counter.count(), 1);
 
         counter.record_suppression(now);
         assert_eq!(counter.count(), 2);
-
-        counter.record_suppression(now);
-        assert_eq!(counter.count(), 3);
     }
 
     #[test]
@@ -304,7 +304,7 @@ mod tests {
 
         counter.record_suppression(now);
         counter.record_suppression(now);
-        assert_eq!(counter.count(), 3);
+        assert_eq!(counter.count(), 2);
 
         counter.reset(now);
         assert_eq!(counter.count(), 0);
@@ -322,7 +322,7 @@ mod tests {
         let summary = SuppressionSummary::from_counter(sig, &counter);
 
         assert_eq!(summary.signature, sig);
-        assert_eq!(summary.count, 2);
+        assert_eq!(summary.count, 1);
         assert!(summary.duration >= Duration::from_millis(10));
     }
 
@@ -336,7 +336,7 @@ mod tests {
         let summary = SuppressionSummary::from_counter(sig, &counter);
         let message = summary.format_message();
 
-        assert!(message.contains("suppressed 2 times"));
+        assert!(message.contains("suppressed 1 times"));
         assert!(message.contains(&sig.to_string()));
     }
 
@@ -351,7 +351,7 @@ mod tests {
             counter.record_suppression(now);
         }
 
-        assert_eq!(counter.count(), 10_001); // 10,000 + 1 initial
+        assert_eq!(counter.count(), 10_000);
     }
 
     #[test]
@@ -378,8 +378,8 @@ mod tests {
             handle.join().unwrap();
         }
 
-        // 10 threads * 100 updates + 1 initial = 1001
-        assert_eq!(counter.count(), 1001);
+        // 10 threads * 100 updates = 1000
+        assert_eq!(counter.count(), 1000);
     }
 
     #[test]
@@ -391,7 +391,7 @@ mod tests {
         // Immediately create summary (same timestamp)
         let summary = SuppressionSummary::from_counter(sig, &counter);
 
-        assert_eq!(summary.count, 1);
+        assert_eq!(summary.count, 0);
         assert!(summary.duration < Duration::from_millis(1));
     }
 
@@ -401,7 +401,7 @@ mod tests {
         let counter = SuppressionCounter::new(now);
 
         counter.record_suppression(now);
-        assert_eq!(counter.count(), 2);
+        assert_eq!(counter.count(), 1);
 
         counter.reset(now);
         assert_eq!(counter.count(), 0);
@@ -440,8 +440,8 @@ mod tests {
         counter1.record_suppression(now);
 
         // counter2 should not be affected
-        assert_eq!(counter1.count(), 2);
-        assert_eq!(counter2.count(), 1);
+        assert_eq!(counter1.count(), 1);
+        assert_eq!(counter2.count(), 0);
     }
 
     #[test]
@@ -535,7 +535,7 @@ mod tests {
             counter.record_suppression(now);
         }
 
-        assert_eq!(counter.count(), 100_001);
+        assert_eq!(counter.count(), 100_000);
     }
 
     #[test]
@@ -579,7 +579,7 @@ mod tests {
         // Should not panic, timestamps should saturate gracefully
         let _first = counter.first_suppressed();
         let _last = counter.last_suppressed();
-        assert!(counter.count() > 1);
+        assert!(counter.count() > 0);
     }
 
     #[test]
@@ -631,7 +631,7 @@ mod tests {
         let counter_clone2 = Arc::clone(&counter);
         let done_clone2 = Arc::clone(&done);
         let reader = thread::spawn(move || {
-            let mut last_count = 1;
+            let mut last_count = 0;
             while !done_clone2.load(Ordering::Acquire) {
                 let count = counter_clone2.count();
                 // Count should never decrease
@@ -644,8 +644,8 @@ mod tests {
         writer.join().unwrap();
         reader.join().unwrap();
 
-        // Final count should be 101 (1 initial + 100 recorded)
-        assert_eq!(counter.count(), 101);
+        // Final count should be 100
+        assert_eq!(counter.count(), 100);
     }
 
     #[test]
@@ -657,7 +657,7 @@ mod tests {
         // Create summary immediately - same timestamp for first and last
         let summary = SuppressionSummary::from_counter(sig, &counter);
 
-        assert_eq!(summary.count, 1);
+        assert_eq!(summary.count, 0);
         assert_eq!(summary.duration, Duration::from_secs(0));
         assert_eq!(summary.first_suppressed, summary.last_suppressed);
     }

--- a/src/infrastructure/layer.rs
+++ b/src/infrastructure/layer.rs
@@ -17,6 +17,8 @@ use crate::infrastructure::storage::ShardedStorage;
 use crate::infrastructure::visitor::FieldVisitor;
 
 use std::borrow::Cow;
+#[cfg(feature = "async")]
+use std::cell::Cell;
 use std::collections::{BTreeMap, BTreeSet};
 use std::sync::Arc;
 use std::time::Duration;
@@ -24,6 +26,33 @@ use tracing::{Metadata, Subscriber};
 use tracing_subscriber::layer::Filter;
 use tracing_subscriber::registry::LookupSpan;
 use tracing_subscriber::{layer::Context, Layer};
+
+#[cfg(feature = "async")]
+thread_local! {
+    /// Guard against re-entrant event processing (e.g., summary emissions triggering throttling).
+    /// Only relevant when active emission is enabled, since summary formatters emit tracing
+    /// events that would otherwise be recursively throttled on the same thread.
+    static IN_THROTTLE: Cell<bool> = const { Cell::new(false) };
+}
+
+/// RAII guard that resets IN_THROTTLE to false on drop, even during panics.
+#[cfg(feature = "async")]
+struct ThrottleGuard;
+
+#[cfg(feature = "async")]
+impl ThrottleGuard {
+    fn enter() -> Self {
+        IN_THROTTLE.with(|flag| flag.set(true));
+        ThrottleGuard
+    }
+}
+
+#[cfg(feature = "async")]
+impl Drop for ThrottleGuard {
+    fn drop(&mut self) {
+        IN_THROTTLE.with(|flag| flag.set(false));
+    }
+}
 
 #[cfg(feature = "async")]
 use crate::application::emitter::{EmitterHandle, SummaryEmitter};
@@ -554,6 +583,7 @@ impl TracingRateLimitLayerBuilder {
 
             let handle = emitter.start(
                 move |summaries| {
+                    let _guard = ThrottleGuard::enter();
                     for summary in summaries {
                         formatter(&summary);
                     }
@@ -906,11 +936,18 @@ where
     }
 
     fn event_enabled(&self, event: &tracing::Event<'_>, cx: &Context<'_, Sub>) -> bool {
+        // Prevent re-entrant throttling (e.g., summary emissions triggering throttling)
+        #[cfg(feature = "async")]
+        if IN_THROTTLE.with(|flag| flag.get()) {
+            return true;
+        }
+
         let metadata_obj = event.metadata();
 
         // Check if this target is exempt from rate limiting
         // Skip the lookup if no exempt targets are configured (common case)
-        if !self.exempt_targets.is_empty() && self.exempt_targets.contains(metadata_obj.target()) {
+        let target = metadata_obj.target();
+        if !self.exempt_targets.is_empty() && self.exempt_targets.contains(target) {
             // Exempt targets bypass rate limiting entirely
             self.limiter.metrics().record_allowed();
             return true;
@@ -1585,5 +1622,48 @@ mod tests {
         assert!(count > 0, "Default formatter should have emitted summaries");
 
         handle.shutdown().await.expect("shutdown failed");
+    }
+
+    #[cfg(feature = "async")]
+    #[tokio::test]
+    async fn test_summary_emission_not_recursively_throttled() {
+        use std::time::Duration;
+
+        // Build a layer with active emission and a very restrictive policy
+        let layer = TracingRateLimitLayer::builder()
+            .with_policy(Policy::count_based(1).unwrap())
+            .with_active_emission(true)
+            .with_summary_interval(Duration::from_millis(100))
+            .build()
+            .unwrap();
+
+        let layer_clone = layer.clone();
+
+        let subscriber = tracing_subscriber::registry()
+            .with(tracing_subscriber::fmt::layer().with_filter(layer));
+
+        // Emit events to trigger suppressions, then wait for multiple summary intervals.
+        // If summaries were recursively throttled, the second interval would produce
+        // nested "Suppressed 1 times: ... Suppressed N times: ..." messages and the
+        // signature count would keep growing.
+        tracing::subscriber::with_default(subscriber, || {
+            // Trigger suppressions
+            for _ in 0..5 {
+                tracing::info!(target: "myapp", "repetitive event");
+            }
+        });
+
+        // Wait for two summary intervals
+        tokio::time::sleep(Duration::from_millis(250)).await;
+
+        // Signature count should be exactly 1 (the "myapp" event).
+        // If summary events were being throttled, additional signatures would appear.
+        assert_eq!(
+            layer_clone.signature_count(),
+            1,
+            "Summary emissions should not create additional throttle signatures"
+        );
+
+        layer_clone.shutdown().await.expect("shutdown failed");
     }
 }

--- a/src/infrastructure/layer.rs
+++ b/src/infrastructure/layer.rs
@@ -17,8 +17,6 @@ use crate::infrastructure::storage::ShardedStorage;
 use crate::infrastructure::visitor::FieldVisitor;
 
 use std::borrow::Cow;
-#[cfg(feature = "async")]
-use std::cell::Cell;
 use std::collections::{BTreeMap, BTreeSet};
 use std::sync::Arc;
 use std::time::Duration;
@@ -27,32 +25,13 @@ use tracing_subscriber::layer::Filter;
 use tracing_subscriber::registry::LookupSpan;
 use tracing_subscriber::{layer::Context, Layer};
 
+/// Internal target used for summary emission events.
+///
+/// Events emitted with this target are automatically exempt from throttling to
+/// prevent recursive suppression of summary messages. This target is intentionally
+/// internal and not part of the public API.
 #[cfg(feature = "async")]
-thread_local! {
-    /// Guard against re-entrant event processing (e.g., summary emissions triggering throttling).
-    /// Only relevant when active emission is enabled, since summary formatters emit tracing
-    /// events that would otherwise be recursively throttled on the same thread.
-    static IN_THROTTLE: Cell<bool> = const { Cell::new(false) };
-}
-
-/// RAII guard that resets IN_THROTTLE to false on drop, even during panics.
-#[cfg(feature = "async")]
-struct ThrottleGuard;
-
-#[cfg(feature = "async")]
-impl ThrottleGuard {
-    fn enter() -> Self {
-        IN_THROTTLE.with(|flag| flag.set(true));
-        ThrottleGuard
-    }
-}
-
-#[cfg(feature = "async")]
-impl Drop for ThrottleGuard {
-    fn drop(&mut self) {
-        IN_THROTTLE.with(|flag| flag.set(false));
-    }
-}
+const SUMMARY_TARGET: &str = "tracing_throttle::summary";
 
 #[cfg(feature = "async")]
 use crate::application::emitter::{EmitterHandle, SummaryEmitter};
@@ -573,6 +552,7 @@ impl TracingRateLimitLayerBuilder {
             let formatter = self.summary_formatter.unwrap_or_else(|| {
                 Arc::new(|summary: &SuppressionSummary| {
                     tracing::warn!(
+                        target: SUMMARY_TARGET,
                         signature = %summary.signature,
                         count = summary.count,
                         "{}",
@@ -583,7 +563,6 @@ impl TracingRateLimitLayerBuilder {
 
             let handle = emitter.start(
                 move |summaries| {
-                    let _guard = ThrottleGuard::enter();
                     for summary in summaries {
                         formatter(&summary);
                     }
@@ -936,17 +915,19 @@ where
     }
 
     fn event_enabled(&self, event: &tracing::Event<'_>, cx: &Context<'_, Sub>) -> bool {
-        // Prevent re-entrant throttling (e.g., summary emissions triggering throttling)
+        let metadata_obj = event.metadata();
+        let target = metadata_obj.target();
+
+        // Exempt our internal summary emission target to prevent recursive throttling.
+        // Only the default formatter uses this target; custom formatters are the user's
+        // responsibility and are intentionally subject to normal throttling rules.
         #[cfg(feature = "async")]
-        if IN_THROTTLE.with(|flag| flag.get()) {
+        if target == SUMMARY_TARGET {
             return true;
         }
 
-        let metadata_obj = event.metadata();
-
         // Check if this target is exempt from rate limiting
         // Skip the lookup if no exempt targets are configured (common case)
-        let target = metadata_obj.target();
         if !self.exempt_targets.is_empty() && self.exempt_targets.contains(target) {
             // Exempt targets bypass rate limiting entirely
             self.limiter.metrics().record_allowed();
@@ -975,7 +956,7 @@ where
             let event_metadata = crate::domain::metadata::EventMetadata::new(
                 metadata_obj.level().as_str().to_string(),
                 message,
-                metadata_obj.target().to_string(),
+                target.to_string(),
                 combined_fields,
             );
 

--- a/tests/redis_storage.rs
+++ b/tests/redis_storage.rs
@@ -61,7 +61,7 @@ async fn test_redis_basic_set_get() {
         sig,
         || EventState::new(policy.clone(), Instant::now()),
         |state| {
-            assert_eq!(state.counter.count(), 1);
+            assert_eq!(state.counter.count(), 0);
         },
     );
 
@@ -70,7 +70,7 @@ async fn test_redis_basic_set_get() {
         sig,
         || panic!("Should not create new state"),
         |state| {
-            assert_eq!(state.counter.count(), 1);
+            assert_eq!(state.counter.count(), 0);
             state.counter.record_suppression(Instant::now());
         },
     );
@@ -80,7 +80,7 @@ async fn test_redis_basic_set_get() {
         sig,
         || panic!("Should not create new state"),
         |state| {
-            assert_eq!(state.counter.count(), 2);
+            assert_eq!(state.counter.count(), 1);
         },
     );
 
@@ -230,7 +230,7 @@ async fn test_redis_concurrent_access() {
     }
 
     // Final count should reflect all increments
-    // Note: Due to Redis race conditions, this might not be exactly 101 (1 initial + 100 records)
+    // Note: Due to Redis race conditions, this might not be exactly 100
     // but should be at least close
     storage.with_entry_mut(
         sig,
@@ -484,13 +484,13 @@ async fn test_redis_with_registry() {
 
     // Use registry to access state
     registry.with_event_state(sig, |state, now| {
-        assert_eq!(state.counter.count(), 1);
+        assert_eq!(state.counter.count(), 0);
         state.counter.record_suppression(now);
     });
 
     // Verify persistence
     registry.with_event_state(sig, |state, _| {
-        assert_eq!(state.counter.count(), 2);
+        assert_eq!(state.counter.count(), 1);
     });
 
     storage.clear();

